File: javascript/LinkedList/LinkedList_CycleDetection.js
import { ListNode } from "./ListNode.js";

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        const map = new Map();
        let cur = head;
        let index = 0;
        while (cur != null) {

            if (map.has(cur)) {
                return map.get(cur);
            }

            map.set(cur, index);
            index++;
            cur = cur.next;
        }

        return -1;
    }
}

const tail = new ListNode(3);
const head = new ListNode(
    0,
    new ListNode(
        1,
        new ListNode(
            2,
            tail
        )
    )
);
tail.next = head;

console.log(
    new Solution().hasCycle(head)
);

File: java/src/com/leetcode/refresh/LinkedListCycleDetection.java
package com.leetcode.refresh;

import com.leetcode.linkedlist.ListNode;

public class LinkedListCycleDetection {
    public static final ListNode TAIL = new ListNode(3);
    public static final ListNode HEAD = new ListNode(0,
            new ListNode(1,
                    new ListNode(2,
                            TAIL)));

    static {
        TAIL.next = HEAD;
    }

    public static void main(String[] args) {
        System.out.println(
                new LinkedListCycleDetection().hasCycle(HEAD)
        );
    }

    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        var slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (fast == slow) {
                return true;
            }

        }


        return false;
    }
}
